### DISCUSSION TITLE
Import Error: No module named 'pyverilator.util'

### DISCUSSION THREAD

**Discussion by ThiamismMouhamed** [Xilinx/finn#1221] [Q&A]
Hello everyone,

I’m encountering an issue while running the **2-import-into-finn-and-verify.ipynb** notebook from the FINN project. Here are the details:

#### Context

- **Setup**: I have downloaded and installed `pyverilator` to meet the requirements for this notebook.
- **Goal**: I’m trying to use the `inference_with_finn_onnx` function to compare predictions between the Brevitas model and the FINN-processed model.
- **FINN Version**: I’m using the latest version from GitHub.

#### Code and Error

The error occurs when I run the following code:
```python
import finn.core.onnx_exec as oxe

def inference_with_finn_onnx(current_inp):
    # Processing code
    output_dict = oxe.execute_onnx(model_for_sim, input_dict)
    return output_dict[finnonnx_out_tensor_name]
```

I receive this error message:

```
ModuleNotFoundError: No module named 'pyverilator.util'
```

Upon examining the installed `pyverilator` package, I confirmed that there is no `util` submodule within it. The `pyverilator` directory contains only the following files:
- `__init__.py`
- `__main__.py`
- `pyverilator.py`
- `verilatorcpp.py`

**Troubleshooting Steps**

- I verified the `pyverilator` installation with `pip show pyverilator`, which confirms that it is installed in my Python environment.
- I reviewed the available documentation and files for `pyverilator`, but couldn’t find any mention of a `util` submodule.

#### Questions

1. Is there a specific version of `pyverilator` that includes the `util` submodule? Or has this module been deprecated or renamed?
2. Could this be a compatibility issue with the current version of FINN?
3. Are there any alternative methods or workarounds to address this issue?

Thank you in advance for any guidance you can provide!

<img width="837" alt="Capture d’écran 2024-10-31 à 15 09 19" src="https://github.com/user-attachments/assets/39ba7f5c-214b-44ce-96a7-06ec38428128">

**Comment by fpjentzsch** [comment]
Hi,

FINN uses this fork of PyVerilator: https://github.com/maltanar/pyverilator

Due to the multitude of dependencies, you should run FINN only within the included Docker container, as per the Getting Started Guide.
